packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -146,169 +150,144 @@ impl PostgresDatabase {
-                created_by
-            ) 
-            VALUES ($1, $2, $3)
-            RETURNING version_id;
-            "#,
-        )
-        .bind(version_id)
-        .bind(
-            serde_json::to_value(property_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize property type: {err}")),
-        )
-        .bind(created_by)
-        .fetch_one(&self.pool)
-        .await
-        .report()
-        .change_context(InsertionError)
-        .attach_lazy(|| property_type.clone())?;
+    /// - If the [`BaseUri`] does not already exist
+    ///
+    /// [`BaseUri`]: crate::types::BaseUri
+    async fn update<T>(
+        &self,
+        database_type: T,
+        updated_by: AccountId,
+    ) -> Result<Qualified<T>, UpdateError>
+    where
+        T: DatabaseType + Serialize + Send + Sync,
+    {
+        let uri = database_type.uri();
 
-        Ok(())
+        if !self
+            .contains_base_uri(uri.base_uri())
+            .await
+            .change_context(UpdateError)?
+        {
+            return Err(Report::new(BaseUriDoesNotExist)
+                .attach_printable(uri.base_uri().clone())
+                .change_context(UpdateError));
+        }
+
+        let version_id = self.insert_uri(uri).await.change_context(UpdateError)?;
+
+        self.insert_with_id(version_id, &database_type, updated_by)
+            .await
+            .change_context(UpdateError)?;
+
+        Ok(Qualified::new(version_id, database_type, updated_by))
     }
 
-    /// Inserts an [`EntityType`] identified by [`VersionId`], and associated with an [`AccountId`],
-    /// into the database.
+    /// Inserts a [`DatabaseType`] identified by [`VersionId`], and associated with an
+    /// [`AccountId`], into the database.
     ///
     /// # Errors
     ///
     /// - [`DatastoreError`], if inserting failed.
-    async fn insert_entity_type(
-        &mut self,
+    async fn insert_with_id<T>(
+        &self,
         version_id: VersionId,
-        entity_type: &EntityType,
+        database_type: &T,
         created_by: AccountId,
-    ) -> Result<(), InsertionError> {
-        sqlx::query_as::<_, (Uuid,)>(
+    ) -> Result<(), InsertionError>
+    where
+        T: DatabaseType + Serialize + Sync,
+    {
+        sqlx::query(&format!(
             r#"
-            INSERT INTO entity_types (
+            INSERT INTO {} (
                 version_id,
                 schema,
                 created_by
             ) 
-            VALUES ($1, $2, $3)
-            RETURNING version_id;
+            VALUES ($1, $2, $3);
             "#,
-        )
+            T::table()
+        ))
         .bind(version_id)
         .bind(
-            serde_json::to_value(entity_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize entity type: {err}")),
+            serde_json::to_value(database_type)
+                .unwrap_or_else(|err| unreachable!("Could not serialize data type: {err}")),
```

Comment:
I don't think this `unreachable` message makes sense.

Reply:
I changed the `unreachable` to an error, just in case this is failing for whatever reason:
> This conversion can fail if `T`'s implementation of `Serialize` decides to fail, or if `T` contains a map with non-string keys.

Reply:
I'm alright with that too. But all I meant is that this isn't a "data type" 😁